uuid.go

```diff
@@ -131,6 +131,32 @@ func TimestampFromV6(u UUID) (Timestamp, error) {
 	return Timestamp(uint64(low) + (uint64(mid) << 12) + (uint64(hi) << 28)), nil
 }
 
+// TimestampFromV7 returns the Timestamp embedded within a V7 UUID. This
+// function returns an error if the UUID is any version other than 7.
+//
+// This is implemented based on revision 03 of the Peabody UUID draft, and may
+// be subject to change pending further revisions. Until the final specification
+// revision is finished, changes required to implement updates to the spec will
+// not be considered a breaking change. They will happen as a minor version
+// releases until the spec is final.
+func TimestampFromV7(u UUID) (Timestamp, error) {
+	if u.Version() != 7 {
+		return 0, fmt.Errorf("uuid: %s is version %d, not version 6", u, u.Version())
+	}
+
+	t := 0 |
```

Comment:
Is this correct?  The spec says the timestamp is the first 36 bits and this is grabbing 40.

> UUIDv7 utilizes a 36-bit big-endian unsigned Unix Timestamp value

Reply:
hmm. The spec says 48bit.
> 48 bit big-endian unsigned number of Unix epoch timestamp
> -- ref: https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#name-uuid-version-7

I mostly just reversed the operations in the generator code from here:
https://github.com/gofrs/uuid/blob/8318aa52a5fd64b188587b18f738ac099d4bcdb6/generator.go#L377-L382

Reply:
You're correct.  I had a link to an older draft.  That being said, both of those snippets of care are working with 40 bits, not 48.

Did they remove support for sub-millisecond granularity?  I haven't been keeping up.

Reply:
> With UUID version 1 and 6, 100-nanoseconds of precision are present while *UUIDv7 features fixed millisecond level of precision* within the Unix epoch that does not exceed the granularity capable in most modern systems. For other levels of precision UUIDv8 SHOULD be utilized.

Looks like they did.  Both of those code blocks need to add a 6th byte.

Reply:
I'm confused about what you are saying is needed here.

> That being said, both of those snippets of code are working with 40 bits, not 48.

The shift is by 40 bits, but 48 bits are being used. (a[0]...a[5])

```
|-- 48bits / 6 bytes --------------------------|
aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff = 281474976710655 (max)
                                        |-byte-|                          bytes 48-40
0000000000000000000000000000000000000000aaaaaaaa = >>40 -> byte([a]>>40) == 'aaaaaaa'
00000000000000000000000000000000aaaaaaaabbbbbbbb = >>32
000000000000000000000000aaaaaaaabbbbbbbbcccccccc = >>24
0000000000000000aaaaaaaabbbbbbbbccccccccdddddddd = >>16
00000000aaaaaaaabbbbbbbbccccccccddddddddeeeeeeee = >>8
aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff = >>0
```

Reply:
You're right again.  I was multi-tasking and messed up the math.